suddenly action fails today

Thanks for making this action! It's quite nice to have.

Yesterday, the action started failing with this trace. I'm not sure if it's something on our side that triggered it or not. We do have a lot of failing tests in this project, which are ignored by a flag on the test action. That was never an issue before. 

<https://github.com/usethesource/rascal-core/runs/6112088376?check_suite_focus=true>

```
 2022-04-21 13:29:23 +0000 - publish-unit-test-results -  INFO - reading target/surefire-reports/**/*.xml
  2022-04-21 13:29:24 +0000 - publish -  INFO - publishing failure results for commit a074b5e8be95a4e6367c6af579f31e768fd77845
  2022-04-21 13:29:26 +0000 - publish -  INFO - created check https://github.com/usethesource/rascal-core/runs/6112893594
  Traceback (most recent call last):
    File "/home/runner/work/_actions/EnricoMi/publish-unit-test-result-action/v1/composite/../python/publish_unit_test_results.py", line 331, in <module>
      main(settings, gha)
    File "/home/runner/work/_actions/EnricoMi/publish-unit-test-result-action/v1/composite/../python/publish_unit_test_results.py", line 102, in main
      Publisher(settings, gh, gha).publish(stats, results.case_results, conclusion)
    File "/home/runner/work/_actions/EnricoMi/publish-unit-test-result-action/v1/python/publish/publisher.py", line 97, in publish
      check_run = self.publish_check(stats, cases, conclusion)
    File "/home/runner/work/_actions/EnricoMi/publish-unit-test-result-action/v1/python/publish/publisher.py", line 289, in publish_check
      check_run.edit(output=output)
    File "/home/runner/.local/lib/python3.8/site-packages/github/CheckRun.py", line 261, in edit
      headers, data = self._requester.requestJsonAndCheck(
    File "/home/runner/.local/lib/python3.8/site-packages/github/Requester.py", line 353, in requestJsonAndCheck
      return self.__check(
    File "/home/runner/.local/lib/python3.8/site-packages/github/Requester.py", line 378, in __check
      raise self.__createException(status, responseHeaders, output)
  github.GithubException.GithubException: 422 {"message": "Validation Failed", "errors": [{"resource": "CheckRun", "code": "invalid", "field": "annotations"}], "documentation_url": "https://docs.github.com/rest/reference/checks#update-a-check-run"}
  Error: Process completed with exit code 1.
```

--- a/python/publish/__init__.py
+++ b/python/publish/__init__.py
@@ -153,7 +153,7 @@ def restrict_unicode(text: Optional[str]) -> Optional[str]:
                     for c in text])
 
 
-def restrict_unicode_list(texts: List[str]) -> List[str]:
+def restrict_unicode_list(texts: List[Optional[str]]) -> List[Optional[str]]:
     return [restrict_unicode(text) for text in texts]
 
 

--- a/python/publish/__init__.py
+++ b/python/publish/__init__.py
@@ -144,6 +144,19 @@ def utf8_character_length(c: int) -> int:
     return 1
 
 
+# Github API does not like Unicode characters above 0xffff
+# Those characters are replaced here by \U00000000
+def restrict_unicode(text: Optional[str]) -> Optional[str]:
+    if text is None:
+        return None
+    return ''.join([r"\U{:08x}".format(ord(c)) if ord(c) > 0xffff else c
+                    for c in text])
+
+
+def restrict_unicode_list(texts: List[str]) -> List[str]:
+    return [restrict_unicode(text) for text in texts]
+
+
 def alternating_range(positive_first: bool = True) -> Iterator[int]:
     i = 0
     yield i
@@ -628,7 +641,7 @@ class Annotation:
         dictionary = self.__dict__.copy()
         dictionary['message'] = abbreviate_bytes(dictionary['message'], 64000)
         dictionary['title'] = abbreviate(dictionary['title'], 255)
-        dictionary['raw_details'] = abbreviate(dictionary['raw_details'], 64000)
+        dictionary['raw_details'] = abbreviate(restrict_unicode(dictionary['raw_details']), 64000)
         if not dictionary.get('start_column'):
             del dictionary['start_column']
         if not dictionary.get('end_column'):
@@ -758,11 +771,11 @@ def get_skipped_tests_list(cases: UnitTestCaseResults) -> List[str]:
 
 
 def get_all_tests_list_annotation(cases: UnitTestCaseResults, max_chunk_size: int = 64000) -> List[Annotation]:
-    return get_test_list_annotation(get_all_tests_list(cases), 'test', max_chunk_size)
+    return get_test_list_annotation(restrict_unicode_list(get_all_tests_list(cases)), 'test', max_chunk_size)
 
 
 def get_skipped_tests_list_annotation(cases: UnitTestCaseResults, max_chunk_size: int = 64000) -> List[Annotation]:
-    return get_test_list_annotation(get_skipped_tests_list(cases), 'skipped test', max_chunk_size)
+    return get_test_list_annotation(restrict_unicode_list(get_skipped_tests_list(cases)), 'skipped test', max_chunk_size)
 
 
 def get_test_list_annotation(tests: List[str], label: str, max_chunk_size: int = 64000) -> List[Annotation]:

--- a/python/publish/publisher.py
+++ b/python/publish/publisher.py
@@ -347,10 +347,10 @@ class Publisher:
                      for test in test_list]
         return test_list
 
-    def get_test_list_annotations(self, cases: UnitTestCaseResults) -> List[Annotation]:
-        all_tests = get_all_tests_list_annotation(cases) \
+    def get_test_list_annotations(self, cases: UnitTestCaseResults, max_chunk_size: int = 64000) -> List[Annotation]:
+        all_tests = get_all_tests_list_annotation(cases, max_chunk_size) \
             if all_tests_list in self._settings.check_run_annotation else []
-        skipped_tests = get_skipped_tests_list_annotation(cases) \
+        skipped_tests = get_skipped_tests_list_annotation(cases, max_chunk_size) \
             if skipped_tests_list in self._settings.check_run_annotation else []
         return [annotation for annotation in skipped_tests + all_tests if annotation]
 

--- a/python/test/files/unsupported-unicode.xml
+++ b/python/test/files/unsupported-unicode.xml
@@ -1,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+<testsuites>
+    <testsuite errors="2" failures="2" skipped="2" name="pytest" tests="7" time="8.61">
+        <testcase file="test/test-1.py" line="1" name="test 1 헴䜝헱홐㣇㿷䔭𒍺𡓿𠄉㦓" time="1.23"></testcase>
+        <testcase file="test/test-2.py" line="2" name="test 2" time="1.23">
+            <skipped message="Some unsupported unicode characters: 헴䜝헱홐㣇㿷䔭𒍺𡓿𠄉㦓">skipped
+            </skipped>
+        </testcase>
+        <testcase file="test/test-3.py" line="3" name="test 3" time="1.23">
+            <skipped message="message">Some unsupported unicode characters: 헴䜝헱홐㣇㿷䔭𒍺𡓿𠄉㦓
+            </skipped>
+        </testcase>
+        <testcase file="test/test-4.py" line="4" name="test 4" time="1.23">
+            <failure message="Some unsupported unicode characters: 헴䜝헱홐㣇㿷䔭𒍺𡓿𠄉㦓">failed
+            </failure>
+        </testcase>
+        <testcase file="test/test-5.py" line="5" name="test 5" time="1.23">
+            <failure message="message">Some unsupported unicode characters: 헴䜝헱홐㣇㿷䔭𒍺𡓿𠄉㦓
+            </failure>
+        </testcase>
+        <testcase file="test/test-6.py" line="6" name="test 6" time="1.23">
+            <error message=">Some unsupported unicode characters: 헴䜝헱홐㣇㿷䔭𒍺𡓿𠄉㦓">error
+            </error>
+        </testcase>
+        <testcase file="test/test-7.py" line="7" name="test 7" time="1.23">
+            <error message="message">Some unsupported unicode characters: 헴䜝헱홐㣇㿷䔭𒍺𡓿𠄉㦓
+            </error>
+        </testcase>
+    </testsuite>
+</testsuites>

--- a/python/test/test_publish.py
+++ b/python/test/test_publish.py
@@ -93,6 +93,30 @@ class PublishTest(unittest.TestCase):
         self.assertIsNone(changes.remaining_and_un_skipped())
         self.assertIsNone(changes.removed_skips())
 
+    def test_restrict_unicode(self):
+        self.assertEqual(None, restrict_unicode(None))
+        self.assertEqual('', restrict_unicode(''))
+
+        # utf8 characters ≤ 0xffff
+        self.assertEqual('…', restrict_unicode('…'))
+        self.assertEqual('abc', restrict_unicode('abc'))
+        self.assertEqual('»»»»»', restrict_unicode('»»»»»'))
+        self.assertEqual('▊▋▌▍▎', restrict_unicode('▊▋▌▍▎'))
+
+        # utf8 characters > 0xffff
+        self.assertEqual(r'\U0001d482\U0001d483\U0001d484', restrict_unicode('𝒂𝒃𝒄'))
+        self.assertEqual(r'헴䜝헱홐㣇㿷䔭\U0001237a\U000214ff\U00020109㦓', restrict_unicode('헴䜝헱홐㣇㿷䔭𒍺𡓿𠄉㦓'))
+
+        # restricting a second time should not alter the result
+        self.assertEqual(None, restrict_unicode(restrict_unicode(None)))
+        self.assertEqual('', restrict_unicode(restrict_unicode('')))
+        self.assertEqual('…', restrict_unicode(restrict_unicode('…')))
+        self.assertEqual('abc', restrict_unicode(restrict_unicode('abc')))
+        self.assertEqual('»»»»»', restrict_unicode(restrict_unicode('»»»»»')))
+        self.assertEqual('▊▋▌▍▎', restrict_unicode(restrict_unicode('▊▋▌▍▎')))
+        self.assertEqual(r'\U0001d482\U0001d483\U0001d484', restrict_unicode(restrict_unicode('𝒂𝒃𝒄')))
+        self.assertEqual(r'헴䜝헱홐㣇㿷䔭\U0001237a\U000214ff\U00020109㦓', restrict_unicode(restrict_unicode('헴䜝헱홐㣇㿷䔭𒍺𡓿𠄉㦓')))
+
     def test_abbreviate_characters(self):
         # None string
         self.assertIsNone(abbreviate(None, 1))
@@ -1427,6 +1451,16 @@ class PublishTest(unittest.TestCase):
         self.assertEqual('message ' * 3999 + 'message…ssage ' + 'message ' * 3999, annotation.to_dict().get('message'))
         self.assertEqual('title - ' * 15 + 'title -…itle - ' + 'title - ' * 15, annotation.to_dict().get('title'))
         self.assertEqual('raw ' * 8000 + '…aw ' + 'raw ' * 7999, annotation.to_dict().get('raw_details'))
+
+    def test_annotation_to_dict_restricted_unicode(self):
+        annotation = Annotation(path='file1', start_line=123, end_line=123, start_column=4, end_column=5, annotation_level='notice', message='result-file1', title='1 out of 6 runs skipped: test1', raw_details='message abc')
+        self.assertEqual(dict(path='file1', start_line=123, end_line=123, start_column=4, end_column=5, annotation_level='notice', message='result-file1', title='1 out of 6 runs skipped: test1', raw_details='message abc'), annotation.to_dict())
+        annotation = Annotation(path='file1', start_line=123, end_line=123, start_column=4, end_column=5, annotation_level='notice', message='result-file1', title='1 out of 6 runs skipped: test1', raw_details='message »»»')
+        self.assertEqual(dict(path='file1', start_line=123, end_line=123, start_column=4, end_column=5, annotation_level='notice', message='result-file1', title='1 out of 6 runs skipped: test1', raw_details='message »»»'), annotation.to_dict())
+        annotation = Annotation(path='file1', start_line=123, end_line=123, start_column=4, end_column=5, annotation_level='notice', message='result-file1', title='1 out of 6 runs skipped: test1', raw_details='message ▊▋▌▍▎')
+        self.assertEqual(dict(path='file1', start_line=123, end_line=123, start_column=4, end_column=5, annotation_level='notice', message='result-file1', title='1 out of 6 runs skipped: test1', raw_details='message ▊▋▌▍▎'), annotation.to_dict())
+        annotation = Annotation(path='file1', start_line=123, end_line=123, start_column=4, end_column=5, annotation_level='notice', message='result-file1', title='1 out of 6 runs skipped: test1', raw_details='message 𝒂𝒃𝒄')
+        self.assertEqual(dict(path='file1', start_line=123, end_line=123, start_column=4, end_column=5, annotation_level='notice', message='result-file1', title='1 out of 6 runs skipped: test1', raw_details='message \\U0001d482\\U0001d483\\U0001d484'), annotation.to_dict())
 
     def test_get_case_annotation(self):
         messages = CaseMessages([

--- a/python/test/test_publisher.py
+++ b/python/test/test_publisher.py
@@ -244,6 +244,44 @@ class TestPublisher(unittest.TestCase):
                       if not call[0].startswith('_logger.')]
         return mock_calls
 
+    def test_get_test_list_annotations(self):
+        cases = UnitTestCaseResults([
+            ((None, 'class', 'test abcd'), {'success': [None]}),
+            ((None, 'class', 'test efgh'), {'skipped': [None]}),
+            ((None, 'class', 'test ijkl'), {'skipped': [None]}),
+        ])
+
+        settings = self.create_settings(check_run_annotation=[all_tests_list, skipped_tests_list])
+        gh = mock.MagicMock()
+        publisher = Publisher(settings, gh, None)
+        annotations = publisher.get_test_list_annotations(cases, max_chunk_size=42)
+
+        self.assertEqual([
+            Annotation(path='.github', start_line=0, end_line=0, start_column=None, end_column=None, annotation_level='notice', message='There are 2 skipped tests, see "Raw output" for the full list of skipped tests.', title='2 skipped tests found', raw_details='class ‑ test efgh\nclass ‑ test ijkl'),
+            Annotation(path='.github', start_line=0, end_line=0, start_column=None, end_column=None, annotation_level='notice', message='There are 3 tests, see "Raw output" for the list of tests 1 to 2.', title='3 tests found (test 1 to 2)', raw_details='class ‑ test abcd\nclass ‑ test efgh'),
+            Annotation(path='.github', start_line=0, end_line=0, start_column=None, end_column=None, annotation_level='notice', message='There are 3 tests, see "Raw output" for the list of tests 3 to 3.', title='3 tests found (test 3 to 3)', raw_details='class ‑ test ijkl')
+        ], annotations)
+
+    def test_get_test_list_annotations_chunked_and_restricted_unicode(self):
+        cases = UnitTestCaseResults([
+            ((None, 'class', 'test 𝒂'), {'success': [None]}),
+            ((None, 'class', 'test 𝒃'), {'skipped': [None]}),
+            ((None, 'class', 'test 𝒄'), {'skipped': [None]}),
+        ])
+
+        settings = self.create_settings(check_run_annotation=[all_tests_list, skipped_tests_list])
+        gh = mock.MagicMock()
+        publisher = Publisher(settings, gh, None)
+        annotations = publisher.get_test_list_annotations(cases, max_chunk_size=42)
+
+        self.assertEqual([
+            Annotation(path='.github', start_line=0, end_line=0, start_column=None, end_column=None, annotation_level='notice', message='There are 2 skipped tests, see "Raw output" for the list of skipped tests 1 to 1.', title='2 skipped tests found (test 1 to 1)', raw_details='class ‑ test \\U0001d483'),
+            Annotation(path='.github', start_line=0, end_line=0, start_column=None, end_column=None, annotation_level='notice', message='There are 2 skipped tests, see "Raw output" for the list of skipped tests 2 to 2.', title='2 skipped tests found (test 2 to 2)', raw_details='class ‑ test \\U0001d484'),
+            Annotation(path='.github', start_line=0, end_line=0, start_column=None, end_column=None, annotation_level='notice', message='There are 3 tests, see "Raw output" for the list of tests 1 to 1.', title='3 tests found (test 1 to 1)', raw_details='class ‑ test \\U0001d482'),
+            Annotation(path='.github', start_line=0, end_line=0, start_column=None, end_column=None, annotation_level='notice', message='There are 3 tests, see "Raw output" for the list of tests 2 to 2.', title='3 tests found (test 2 to 2)', raw_details='class ‑ test \\U0001d483'),
+            Annotation(path='.github', start_line=0, end_line=0, start_column=None, end_column=None, annotation_level='notice', message='There are 3 tests, see "Raw output" for the list of tests 3 to 3.', title='3 tests found (test 3 to 3)', raw_details='class ‑ test \\U0001d484')
+        ], annotations)
+
     def test_publish_without_comment(self):
         settings = self.create_settings(comment_mode=comment_mode_off, hide_comment_mode=hide_comments_mode_off)
         mock_calls = self.call_mocked_publish(settings, prs=[object()])

